### PR TITLE
feat: decode Optimism blobs

### PR DESCRIPTION
**Pull request by PabloCastellano**
### Checklist

- [ ] My change requires a documentation update, and I have done it.
- [ ] I have added tests to cover my changes.
- [x] I have filled out the description and linked the related issues.

### Description

Add initial batches decoder for [Ecotone](https://docs.optimism.io/builders/node-operators/network-upgrades) network upgrade.

Note: this will only work for blobs sent between Thu Mar 14 00:00:01 UTC 2024 (1710374401) and Wed Jul 10 16:00:01 UTC 2024 (1720627201). **Out of 36207 transactions, we could decode most of them (36039) but 168 failed (0,4%). This will be addressed in a different PR.**

Supports decoding the following fields from the combination of blobs:

* Timestamp since L2 genesis
* Last L1 origin number
* Parent L2 block hash
* L1 origin block hash
* Number of L2 blocks
* How many were changed by L1 origin
* Total txs
* Contract creation txs number

### Package @blobscan/optimism-decoder

The following commands are provided:

**pnpm op:download**

This command will download blobs from Blobscan API for all Optimism Ecotone transactions and will store them in `./blobs/tx`. Blobs will be appended and the final file will be decoded, saving results to `./blobs/tx/<txId>/decoded.json`.

Note: this command requires the file `optimism_ecotone.txt`, which can be generated using the `op:generate_ecotone_txt` command or directly from the database like:

`SELECT hash FROM transaction WHERE from_id = '0x6887246668a3b87f54deb3b94ba47a6f63f32985' AND block_number BETWEEN 19429617 AND 20277188 ORDER BY block_number;`

**pnpm op:generate_ecotone_txt**

Retrieves Ecotone upgrade transactions from database and writes them to `optimism_ecotone.txt`. These are transactions from sender `0x6887246668a3b87f54deb3b94ba47a6f63f32985` and blocks 19429617 to 20277188.

**npm run op:decode**

Decode the information contained in a channel retrieved from multiple blobs. 

Example usage:

```
$ pnpm op:decode ./opstack_blobs_19538908.bin

Timestamp since L2 genesis: 25636980
Last L1 origin number: 19538899
Parent L2 block hash: 04f77f137bf3d52e7fe827254134b0540bc9206e
L1 origin block hash: 297e7aabd7deb7db3475f56dcecf3bb3dca6aefb
Number of L2 blocks: 138
How many were changed by L1 origin: 22
Total txs: 2713
Contract creation txs number: 0
```

**pnpm op:load-decoded-data-to-db**

Example usage: `npm run op:load-decoded-data-to-db ./src/blobs/tx`

This will traverse the directory and check if decoded.json exists. In this case, will update the new `transaction.decoded_fields` column for the transactions.

**pnpm op:generate-import-csv**

This will traverse the directory create a csv file with two columns: transaction hash and decoded_fields. This csv can later be imported using SQL like:

```
CREATE TABLE transaction_tmp(                                                                           
    hash TEXT PRIMARY KEY,                                                                                             
    decoded_fields JSONB                                                                                                
);
\COPY transaction_tmp FROM '/tmp/ecotone_decoded_fields.csv' DELIMITER ','  CSV HEADER;                  
UPDATE transaction SET decoded_fields = tmp.decoded_fields FROM transaction_tmp tmp WHERE transaction.hash = tmp.hash;
DROP TABLE transaction_tmp;
```

### Related Issue (Optional)
#255 

### Screenshots (if appropriate):

image

### CODE REVIEW

File: packages/optimism-decoder/src/decoder.ts [hunk 1]
```diff
@@ -0,0 +1,261 @@
...
+      const tailA = chunk.slice(32 * 0 + 1, 32 * 1);
+      const tailB = chunk.slice(32 * 1 + 1, 32 * 2);
+      const tailC = chunk.slice(32 * 2 + 1, 32 * 3);
+      const tailD = chunk.slice(32 * 3 + 1, 32 * 4);
+
+      const x = (byteA & 0b0011_1111) | ((byteB & 0b0011_0000) << 2);
+      const y = (byteB & 0b0000_1111) | ((byteD & 0b0000_1111) << 4);
+      const z = (byteC & 0b0011_1111) | ((byteD & 0b0011_0000) << 2);
+
+      const result = new Uint8Array(4 * 31 + 3);
```
Comment (xFJA) [inline packages/optimism-decoder/src/decoder.ts:76]: I would avoid magic numbers. We could create a const with a meaningful name. Also, it's repeated in `L85`)

File: packages/optimism-decoder/src/decoder.ts [hunk 2]
```diff
@@ -0,0 +1,261 @@
...
+    lastL1OriginNumber: l1OriginNumberResult.value,
+    parentL2BlockHash: parentL2BlockHashResult.hex,
+    l1OriginBlockHash: l1OriginBlockHashResult.hex,
+    numberOfL2Blocks: l2BlocksNumber,
+    changedByL1Origin: originChangesCount,
+    totalTxs: totalTxs,
+    contractCreationTxsNumber: contractCreationTxsNumber,
+  } as OptimismDecodedData;
+
+  /*
```
Comment (xFJA) [inline packages/optimism-decoder/src/decoder.ts:163]: Debug code? should we delete it?

File: packages/optimism-decoder/src/index.ts [hunk 3]
```diff
@@ -0,0 +1,31 @@
+import { decodeOptimismFile } from "./decoder";
+
+export { decodeOptimismFile } from "./decoder";
+
+const args = process.argv.slice(2); // Slice to get arguments after 'node' and script path
+const filename = args[0]; // The first argument should be the filename
+// const filename = "opstack_blobs_19538908.bin";
```
Comment (xFJA) [inline packages/optimism-decoder/src/index.ts:7]: Debugging code to delete?

File: packages/optimism-decoder/src/index.ts [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+decodeOptimismFile(filename).then((data) => {
+  console.log("Timestamp since L2 genesis:", data.timestampSinceL2Genesis);
+  console.log("Last L1 origin number:", data.lastL1OriginNumber);
+  console.log("Parent L2 block hash:", data.parentL2BlockHash);
+  console.log("L1 origin block hash:", data.l1OriginBlockHash);
+  console.log("Number of L2 blocks:", data.numberOfL2Blocks);
+  console.log("How many were changed by L1 origin:", data.changedByL1Origin);
+  console.log("Total txs:", data.totalTxs);
+  console.log("Contract creation txs number:", data.contractCreationTxsNumber);
+  /*
```
Comment (xFJA) [inline packages/optimism-decoder/src/index.ts:23]: Debugging code to delete?

File: packages/api/src/routers/indexer/indexData.ts [hunk 5]
```diff
@@ -120,6 +120,13 @@ export const indexData = jwtAuthedProcedure
       const dbBlobs = createDBBlobs(input);
       const dbBlobsOnTransactions = createDBBlobsOnTransactions(input);
 
+      /*
+      TODO: Once we support Optimism Fjord network upgrade
```
[the remark this answers is not in the document]
Comment (PabloCastellano) [reply packages/api/src/routers/indexer/indexData.ts:124]: Just added it to https://github.com/Blobscan/blobscan/issues/255

File: packages/optimism-decoder/src/decoder.ts (hunk 1 above)
Comment (PabloCastellano) [reply packages/optimism-decoder/src/decoder.ts:78]: these numbers come from the specification

File: packages/optimism-decoder/src/decoder.ts (hunk 2 above)
Comment (PabloCastellano) [reply packages/optimism-decoder/src/decoder.ts:165]: I prefer to keep it. This is the implementation to decode more fields. However it is not working and I decided to leave it commented. I've added some tasks to https://github.com/Blobscan/blobscan/issues/255 to keep track of it

**Comment by luis-herasme** [comment]
image

**Pull request merged**